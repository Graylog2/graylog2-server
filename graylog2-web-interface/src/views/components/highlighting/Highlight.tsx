import * as React from 'react';
import styled, { css } from 'styled-components';
import type { DefaultTheme } from 'styled-components';

type BackgroundColorProps = {
  theme: DefaultTheme,
  $color: string,
};

const BackgroundColor = styled.div(({ theme, $color }: BackgroundColorProps) => css`
  background-color: ${$color};
  color: ${theme.utils.contrastingColor($color)};
  width: fit-content;
`);

type Props = {
  color: string,
};

const Highlight = ({ children, color }: React.PropsWithChildren<Props>) => <BackgroundColor $color={color}>{children}</BackgroundColor>;

export default Highlight;
