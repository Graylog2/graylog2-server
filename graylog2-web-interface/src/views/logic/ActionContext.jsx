// @flow strict
import * as React from 'react';

export type ActionContexts = { [string]: * };

const ActionContext = React.createContext<ActionContexts>({});

type Props = {
  children: React.Node,
  value: ActionContexts;
};

const AdditionalContext = {
  Provider: ({ children, value }: Props) => (
    <ActionContext.Consumer>
      {(contexts) => (
        <ActionContext.Provider value={{ ...contexts, ...value }}>
          {children}
        </ActionContext.Provider>
      )}
    </ActionContext.Consumer>
  ),
  Consumer: ActionContext.Consumer,
};

export { ActionContext, AdditionalContext };
