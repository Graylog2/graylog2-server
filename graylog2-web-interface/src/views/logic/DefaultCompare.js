// @flow strict
/*
 * Default (string) comparison function used for sorting lists/arrays.
 *
 * This module should be used whenever a consistent, high-performance comparison function for sorting is required. It is
 * especially useful when used instead of `naturalSort` to avoid its excessive usage of regular expressions, but with the
 * `{ numeric: true }` option, a subset (and probably the most desired one) of its functionality can be replaced while
 * being magnitudes faster.
 *
 * The `compare` function returns the comparison function, so it needs to be called before its result is passed to e.g.
 * `sort()`. This is because it instantiates a `Intl.Collator` instance, which is reused in subsequent calls.
 * If only the default options (non-numeric, case-insensitive) are required, `defaultCompare` can be used which does not
 * require to be called before and is therefore a drop-in replacement for `naturalSort`.
 */

/*
 * This generator function can be used to instantiate a comparison method with non-standard options. The usable options
 * are listed here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Collator -> Syntax
 * -> Parameters -> options
 *
 * The `undefined` parameter when instantiating triggers using the runtime's default locales (i.e. the locales supplied
 * by the browser).
 */

export const compare = (options: * = {}) => new Intl.Collator(undefined, options).compare;

/*
 * This is the default comparison function. It compares non-numerically and case-insensitive. It can be used as a drop-in
 * replacement for `naturalSort` and is shared across its consumers. If the default options do not match, `compare` must
 * be used instead.
 */
export const defaultCompare: (any, any) => number = compare();
