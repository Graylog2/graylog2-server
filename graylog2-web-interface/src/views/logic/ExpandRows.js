/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { flatten, setWith } from 'lodash';

const expandRows = (fieldNames, columnFieldNames, rows) => {
  if (!rows) {
    return [];
  }

  const expanded = [];

  rows.forEach((row) => {
    const { values } = row;
    const result = {};

    row.key.forEach((key, idx) => {
      result[fieldNames[idx]] = key;
    });

    values.forEach(({ key, value }) => {
      const translatedKeys = flatten(key.map((k, idx) => (idx < key.length - 1 && columnFieldNames[idx] ? [columnFieldNames[idx], k] : k)));

      setWith(result, translatedKeys, value, Object);
    });

    expanded.push(result);
  });

  return expanded;
};

export default expandRows;
