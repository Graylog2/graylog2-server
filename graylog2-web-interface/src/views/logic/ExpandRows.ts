/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import flatten from 'lodash/flatten';
import setWith from 'lodash/setWith';

import type { ColLeaf, Leaf, NonLeaf, RowInner, RowLeaf } from 'views/logic/searchtypes/pivot/PivotHandler';

type ResultEntry = Result | string | number;

type Result = { [key: string]: ResultEntry };
type Results = Array<Result>;

const expandRows = (fieldNames: Array<string>, columnFieldNames: Array<string>, rows: Array<Leaf | NonLeaf>): Results => {
  if (!rows) {
    return [];
  }

  const expanded = [];

  rows.forEach((row: Leaf | NonLeaf) => {
    const { values } = row;
    const result = {};

    row.key.forEach((key, idx) => {
      result[fieldNames[idx]] = key;
    });

    (values as Array<ColLeaf | RowLeaf | RowInner>).forEach(({ key, value }) => {
      const translatedKeys = flatten(key.map((k, idx) => (idx < key.length - 1 && columnFieldNames[idx] ? [columnFieldNames[idx], k] : k)));

      setWith(result, translatedKeys, value, Object);
    });

    expanded.push(result);
  });

  return expanded;
};

export default expandRows;
