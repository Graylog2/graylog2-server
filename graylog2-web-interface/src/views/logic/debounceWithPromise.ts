import { debounce } from 'lodash';

type PromiseReturnType<T> = T extends (...args: any[]) => Promise<infer R> ? R : never;

const debouncedValidateQuery = <T extends (...args: any[]) => Promise<any>>(fn: T, delay: number) => {
  const debouncedFn = debounce((resolve: PromiseReturnType<T>, ...args: Parameters<T>) => fn(...args).then(resolve), delay);

  return (...args: Parameters<T>) => new Promise<PromiseReturnType<T>>((resolve: PromiseReturnType<T>) => {
    debouncedFn(resolve, ...args);
  });
};

export default debouncedValidateQuery;
