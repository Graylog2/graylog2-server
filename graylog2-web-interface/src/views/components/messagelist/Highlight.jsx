// @flow strict
import * as React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import { AdditionalContext } from 'views/logic/ActionContext';
import { DEFAULT_HIGHLIGHT_COLOR } from 'views/Constants';
import PossiblyHighlight from './PossiblyHighlight';

type Props = {
  field: string,
  value: any,
};

const Highlight = ({ field, value }: Props) => (
  <AdditionalContext.Consumer>
    {({ message }) => (
      <PossiblyHighlight field={field}
                         color={DEFAULT_HIGHLIGHT_COLOR}
                         value={value}
                         highlightRanges={get(message, 'highlight_ranges')} />
    )}
  </AdditionalContext.Consumer>
);

Highlight.propTypes = {
  field: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired,
};

export default Highlight;
