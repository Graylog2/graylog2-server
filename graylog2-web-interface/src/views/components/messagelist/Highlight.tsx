/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import { AdditionalContext } from 'views/logic/ActionContext';
import { DEFAULT_HIGHLIGHT_COLOR } from 'views/Constants';

import PossiblyHighlight from './PossiblyHighlight';

type Props = {
  field: string,
  value: any,
};

const Highlight = ({ field, value }: Props) => (
  <AdditionalContext.Consumer>
    {({ message }) => (
      <PossiblyHighlight field={field}
                         color={DEFAULT_HIGHLIGHT_COLOR}
                         value={value}
                         highlightRanges={get(message, 'highlight_ranges')} />
    )}
  </AdditionalContext.Consumer>
);

Highlight.propTypes = {
  field: PropTypes.string.isRequired,
  value: PropTypes.any.isRequired,
};

export default Highlight;
