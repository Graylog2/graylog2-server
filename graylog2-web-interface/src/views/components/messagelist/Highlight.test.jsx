// @flow strict
import React from 'react';
import { mount } from 'wrappedEnzyme';

import { AdditionalContext } from 'views/logic/ActionContext';
import Highlight from './Highlight';

const messageFor = ranges => ({ highlight_ranges: ranges });
describe('Highlight', () => {
  it('works for empty field & value', () => {
    const wrapper = mount(<Highlight field="" value="" />);
    expect(wrapper).toMatchSnapshot();
  });
  it('returns unmodified string without ranges', () => {
    const wrapper = mount(<Highlight field="foo" value="bar" />);
    expect(wrapper).toMatchSnapshot();
  });
  it('does not highlight string if range for field is absent', () => {
    const message = messageFor({
      bar: [{ start: 0, length: 6 }],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="foobar" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('highlights string for single highlight range', () => {
    const message = messageFor({
      foo: [{ start: 0, length: 6 }],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="foobar" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('does not highlight string if start is negative', () => {
    const message = messageFor({
      foo: [{ start: -3, length: 3 }],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="foobar" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('does not highlight string if length is negative', () => {
    const message = messageFor({
      foo: [{ start: 3, length: -3 }],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="foobar" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('highlights remainder of string if length of range exceeds length of string', () => {
    const message = messageFor({
      foo: [{ start: 3, length: 256 }],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="foobar" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('highlights string for multiple highlight ranges', () => {
    const message = messageFor({
      foo: [
        { start: 4, length: 5 },
        { start: 14, length: 10 },
      ],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="the brown fox jumps over the lazy dog" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('highlights string for multiple, overlapping highlight ranges', () => {
    const message = messageFor({
      foo: [
        { start: 4, length: 5 },
        { start: 7, length: 10 },
      ],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="the brown fox jumps over the lazy dog" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
  it('highlights string for multiple highlight ranges where one is a complete subset of the other', () => {
    const message = messageFor({
      foo: [
        { start: 4, length: 5 },
        { start: 7, length: 1 },
      ],
    });
    const wrapper = mount(
      <AdditionalContext.Provider value={{ message }}>
        <Highlight field="foo" value="the brown fox jumps over the lazy dog" />
      </AdditionalContext.Provider>,
    );
    expect(wrapper.find('PossiblyHighlight')).toMatchSnapshot();
  });
});
