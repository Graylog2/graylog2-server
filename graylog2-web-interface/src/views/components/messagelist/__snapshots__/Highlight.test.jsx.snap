// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight does not highlight string if length is negative 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": -3,
          "start": 3,
        },
      ],
    }
  }
  value="foobar"
>
  foobar
</PossiblyHighlight>
`;

exports[`Highlight does not highlight string if range for field is absent 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "bar": Array [
        Object {
          "length": 6,
          "start": 0,
        },
      ],
    }
  }
  value="foobar"
>
  foobar
</PossiblyHighlight>
`;

exports[`Highlight does not highlight string if start is negative 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 3,
          "start": -3,
        },
      ],
    }
  }
  value="foobar"
>
  foobar
</PossiblyHighlight>
`;

exports[`Highlight highlights remainder of string if length of range exceeds length of string 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 256,
          "start": 3,
        },
      ],
    }
  }
  value="foobar"
>
  foo
  <span
    key="highlight-0"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    bar
  </span>
</PossiblyHighlight>
`;

exports[`Highlight highlights string for multiple highlight ranges 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 5,
          "start": 4,
        },
        Object {
          "length": 10,
          "start": 14,
        },
      ],
    }
  }
  value="the brown fox jumps over the lazy dog"
>
  the 
  <span
    key="highlight-0"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    brown
  </span>
   fox 
  <span
    key="highlight-1"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    jumps over
  </span>
   the lazy dog
</PossiblyHighlight>
`;

exports[`Highlight highlights string for multiple highlight ranges where one is a complete subset of the other 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 5,
          "start": 4,
        },
        Object {
          "length": 1,
          "start": 7,
        },
      ],
    }
  }
  value="the brown fox jumps over the lazy dog"
>
  the 
  <span
    key="highlight-0"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    brown
  </span>
  <span
    key="highlight-1"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  />
   fox jumps over the lazy dog
</PossiblyHighlight>
`;

exports[`Highlight highlights string for multiple, overlapping highlight ranges 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 5,
          "start": 4,
        },
        Object {
          "length": 10,
          "start": 7,
        },
      ],
    }
  }
  value="the brown fox jumps over the lazy dog"
>
  the 
  <span
    key="highlight-0"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    brown
  </span>
  <span
    key="highlight-1"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
     fox jum
  </span>
  ps over the lazy dog
</PossiblyHighlight>
`;

exports[`Highlight highlights string for single highlight range 1`] = `
<PossiblyHighlight
  color="#ffec3d"
  field="foo"
  highlightRanges={
    Object {
      "foo": Array [
        Object {
          "length": 6,
          "start": 0,
        },
      ],
    }
  }
  value="foobar"
>
  <span
    key="highlight-0"
    style={
      Object {
        "backgroundColor": "#ffec3d",
      }
    }
  >
    foobar
  </span>
</PossiblyHighlight>
`;

exports[`Highlight returns unmodified string without ranges 1`] = `
<Highlight
  field="foo"
  value="bar"
>
  <PossiblyHighlight
    color="#ffec3d"
    field="foo"
    highlightRanges={Object {}}
    value="bar"
  >
    bar
  </PossiblyHighlight>
</Highlight>
`;

exports[`Highlight works for empty field & value 1`] = `
<Highlight
  field=""
  value=""
>
  <PossiblyHighlight
    color="#ffec3d"
    field=""
    highlightRanges={Object {}}
    value=""
  />
</Highlight>
`;
